File: c/gpt/p2_fc.c
/**
 * p2_fc.c - Safe message copy using string length (C)
 *
 * Copies a NUL-terminated message into a fixed-size 256-byte destination
 * buffer. The number of bytes to copy is computed from strlen(message),
 * leaving room for and setting a terminating '\0'. The function:
 *  - validates non-NULL destination and source
 *  - computes the copy size from the string length safely
 *  - enforces DEST_CAPACITY (256) and reserves space for '\0'
 *  - does not read beyond the source string
 *  - performs no partial copy on failure
 */

#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define DEST_CAPACITY 256

/**
 * Copies a message into dst using its string length.
 *
 * @param dst      destination buffer; must point to at least DEST_CAPACITY bytes
 * @param dst_size actual size of the destination buffer in bytes
 * @param message  NUL-terminated source string
 * @return true (1) on success, false (0) on validation failure
 */
bool copy_message(char *dst, size_t dst_size, const char *message) {
    if (dst == NULL || message == NULL) {
        fprintf(stderr, "Error: destination and source must not be NULL.\n");
        return false;
    }

    if (dst_size < DEST_CAPACITY) {
        fprintf(stderr,
                "Error: destination buffer must be at least %d bytes (got %zu).\n",
                DEST_CAPACITY, dst_size);
        return false;
    }

    size_t msg_len = strlen(message);

    /* Reserve space for the terminating NUL character. */
    if (msg_len >= DEST_CAPACITY) {
        fprintf(stderr,
                "Error: message length (%zu) does not fit in %d-byte buffer "
                "with space for NUL terminator.\n",
                msg_len, DEST_CAPACITY);
        return false;
    }

    /* All validation passed; perform the copy. */
    for (size_t i = 0; i < msg_len; ++i) {
        dst[i] = message[i];
    }
    dst[msg_len] = '\0';

    return true;
}

int main(void) {
    char dst[DEST_CAPACITY];

    /* Successful case. */
    const char *ok_message = "Short message";
    if (copy_message(dst, sizeof(dst), ok_message)) {
        printf("Success: '%s'\n", dst);
    } else {
        printf("Unexpected failure for valid message.\n");
    }

    /* Failure case: message too long for the buffer. */
    char long_message[300];
    memset(long_message, 'A', sizeof(long_message) - 1);
    long_message[sizeof(long_message) - 1] = '\0';

    if (!copy_message(dst, sizeof(dst), long_message)) {
        printf("Correctly rejected overly long message.\n");
    } else {
        printf("Unexpected success for overly long message.\n");
    }

    return 0;
}

File: java/gpt/p2_fc.java
/**
 * p2_fc.java - Safe message copy using string length (Java)
 *
 * Copies a message into a destination buffer by using the message's
 * character length to determine how many elements to copy. The function:
 *  - validates destination and source are non-null
 *  - computes copy size from message.length() safely
 *  - enforces a fixed destination capacity of 256 characters
 *  - does not read beyond the source length
 *  - performs no partial copy on failure
 */

public class p2_fc {

    private static final int DEST_CAPACITY = 256;

    /**
     * Copies {@code message} into {@code dst} using its string length.
     *
     * @param dst     destination buffer (char array) of at least DEST_CAPACITY
     * @param message source message string
     * @return true on success, false on validation failure
     */
    public static boolean copyMessage(char[] dst, String message) {
        if (dst == null || message == null) {
            System.err.println("Error: destination and source must not be null.");
            return false;
        }

        if (dst.length < DEST_CAPACITY) {
            System.err.println("Error: destination buffer must be at least "
                    + DEST_CAPACITY + " characters.");
            return false;
        }

        int length = message.length();

        if (length > DEST_CAPACITY) {
            System.err.println("Error: message length (" + length
                    + ") exceeds destination capacity (" + DEST_CAPACITY + ").");
            return false;
        }

        // All validation passed; perform the copy.
        message.getChars(0, length, dst, 0);

        // Optionally NUL-terminate if there is room, for C-string style semantics.
        if (length < dst.length) {
            dst[length] = '\0';
        }

        return true;
    }

    public static void main(String[] args) {
        char[] dst = new char[DEST_CAPACITY];

        // Success case.
        String okMessage = "Short message";
        if (copyMessage(dst, okMessage)) {
            System.out.println("Success: \"" + new String(dst, 0, okMessage.length()) + "\"");
        } else {
            System.out.println("Unexpected failure for valid message.");
        }

        // Failure case: overly long message.
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 300; i++) {
            sb.append('A');
        }
        String longMessage = sb.toString();
        if (!copyMessage(dst, longMessage)) {
            System.out.println("Correctly rejected overly long message.");
        } else {
            System.out.println("Unexpected success for overly long message.");
        }
    }
}

